src/core/namespace.c

```diff
@@ -1955,12 +1970,20 @@ static int apply_mounts(
                 bool again = false;
 
                 FOREACH_ARRAY(m, ml->mounts, ml->n_mounts) {
+                        bool chase_inside_root = true;
 
                         if (m->state != MOUNT_PENDING)
                                 continue;
 
-                        /* ExtensionImages/Directories are first opened in the propagate directory, not in the root_directory */
-                        r = follow_symlink(!IN_SET(m->mode, MOUNT_EXTENSION_IMAGE, MOUNT_EXTENSION_DIRECTORY) ? root : NULL, m);
+                        /* ExtensionImages/Directories are first opened in the propagate directory, not in
+                         * the root_directory. A private (invisible to the guest) tmpfs instance is mounted
+                         * on /run/[user/xyz/]systemd/unit-private-tmp as the storage backend of private
+                         * /tmp and /var/tmp. */
+                        if (IN_SET(m->mode, MOUNT_EXTENSION_IMAGE, MOUNT_EXTENSION_DIRECTORY) ||
+                            (m->mode == MOUNT_TMPFS && endswith(mount_entry_path(m), "/systemd/unit-private-tmp")))
+                                chase_inside_root = false;
```

Comment:
Here too, please condition by explicit enums for this, not by path prefix match. Also instead of:

bool b = true;
if (x)
        b = false;

Just do:

b = !x;